tools/jit/gen_unboxing.py

```diff
@@ -0,0 +1,182 @@
+# Generates RegisterCodegenUnboxedKernels.cpp, UnboxingFunctions.h and UnboxingFunctions.cpp.
+import argparse
+import json
+import os
+import pathlib
+from dataclasses import dataclass
+from tools.codegen.api.translate import translate
+from tools.codegen.api.types import CppSignatureGroup
+from tools.codegen.api.unboxing import convert_arguments
+from tools.codegen.context import method_with_native_function
+from tools.codegen.gen import parse_native_yaml
+from tools.codegen.model import NativeFunction, NativeFunctionsGroup, Variant
+from tools.codegen.utils import Target, FileManager, mapMaybe, make_file_manager
+from typing import Union, Sequence
+from typing_extensions import Literal
+
+
+# Generates UnboxingFunctions.h & UnboxingFunctions.cpp.
+@dataclass(frozen=True)
+class ComputeUnboxingFunctions:
```

Comment:
Is there any point in generating the unboxing:: function? It seems to only be called from one site (the OperatorGenerator invocations) and it seems like life would be easier without having to generate these files.

Reply:
That was my suggestion too - one nice thing that lets us do is keep all of the unboxing code under the `at::` (and confirms that we don't accidentally use anything in `torch::` to implement it), which would make it a lot easier to re-use the unboxing codegen in core later if we want to.

There's a separate, mobile-only generated file that generates registrations to the JIT registry. Which we only want to build in mobile, since otherwise we would redundantly register stuff to JIT from both that codegen'd file, and from the dispatcher callback registration logic.

We could in theory just remove the callback logic and have the codegen output be the only place that aten ops get registered - we actually brought that up in one of the Friday syncs, since that would also give this codegen server-side torchscript coverage. But the conclusion was that the main downside of doing so is that we have no way to easily enforce that ops registered to the dispatcher under the `aten` namespace *have* to go through codegen. That conversation might be worth revisiting though.

Separate from that discussion- the extra files are a hassle, so I totally get it if you think it's worth keeping both the unboxing logic and JIT registrations in one file.

Reply:
The layering justification is legitimate; I think my trouble is that I am having trouble thinking of situations where `at::unboxing::` is a useful API, outside of being an API for precisely how it is being used here. A symbol like `at::unboxing::add` is only useful when you know exactly what operator you want (for the JIT interpreter, you need some sort of registry as seen here), and if you know exactly what operator you want, you probably want the unboxed API anyway.

> We could in theory just remove the callback logic and have the codegen output be the only place that aten ops get registered

I'm not sure I follow this.

Reply:
> We could in theory just remove the callback logic and have the codegen output be the only place that aten ops get registered

This is the thing that we talked a bit about yesterday over VC. We want to avoid registering an op to JIT twice, which seems pretty tricky (and the way that the PR deals with it right now is that it only runs the unboxing codegen when static dispatch is enabled, so we don't generate any dispatcher registrations).

One potential way to do that would be to establish some namespace convention, where e.g. all ops in the `aten` namespace are expected to be registered through codegen while everything else i expected to be registered through the callback. It seems kind of hard to check for that invariant though - there's nothing stopping someone from manually registering an op to the dispatcher under the "aten" namespace outside of the codegen, which feels pretty fragile (I dunno, maybe we could route the codegen'd dispatcher registrations through a different, private API? I'm not sure how much effort it's worth putting into this though).